### PR TITLE
Homework6

### PR DESCRIPTION
**Pull request by meszenab**

### CODE REVIEW

File: course5/homework6/string.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#include <memory>
```
Comment (maciekgajewski) [inline course5/homework6/string.h:1]: Begin every header file with ```#pragma once```. Make it your habit

File: course5/homework6/string.cc [hunk 2]
```diff
@@ -0,0 +1,107 @@
+#include <memory>
+#include <cstring>
+#include "string.h"
```
Comment (maciekgajewski) [inline course5/homework6/string.cc:3]: Always order your includes from the most specific ones (local project) to the most general ones (boost, stl)

File: course5/homework6/string.cc [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    {
+        len = src.length();
+        string_pointer = std::make_unique<char[]>(len +1);
+        std::strcpy(string_pointer.get(), src.string_pointer.get());
+    }
+
+    string::string(string&& src)
+    {
+        len = src.length();
+        string_pointer = std::make_unique<char[]>(len +1);
```
Comment (maciekgajewski) [inline course5/homework6/string.cc:26]: You are allocating a buffer here. What will happen with the buffer in the next line?

File: course5/homework6/string.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    public:
+            string();
+            string(const string&);
+            string(string&&);
+            string(const char*);
+            string& operator=(const string&);
+            string& operator=(const char*);
+            string& operator=(string&&);
+            int length() const;
+            char* getchars() const;
```
Comment (maciekgajewski) [inline course5/homework6/string.h:16]: What is this method for? I can't see it being used/implemented?

File: course5/homework6/main.cc [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    s1[2] = 'p';
+    assert( s1[2] == 'p');
+    course::string s3;
+    s3 = s1 + s2;
+    assert( s3 == "wopldhello");
+    s3 += s1;
+    assert( s3 == "wopldhellowopld");
+    std::cout << s3.c_str() << std::endl;
+    s1 = std::move(s3);
+    assert( s1 == "wopldhellowopld");
```
Comment (maciekgajewski) [inline course5/homework6/main.cc:34]: This test is insufficient, leaves mane scenarios/corner cases uncovered.
- Add test of operation on const object. Test all the combination which you'd expect to work with const objects.
- Cover corner cases. Test all the methods on empty string, or with an empty string as an operand.

**Review by maciekgajewski — CHANGES_REQUESTED** [review]
- You are making unnecessary allocations
- Test is incomplete.
- Have you run it past valgrind?

File: course5/homework6/string.h [hunk 6]
```diff
@@ -21,6 +21,18 @@ namespace course{
             bool operator==(const string&);
             bool operator==(const char*);
             char* c_str() const;
+            char* begin();
+            char* begin() const;
+            char* end();
+            char* end() const;
+//            class iterator
```
Comment (maciekgajewski) [inline course5/homework6/string.h:28]: Remove the commented-out code

**Review by maciekgajewski — COMMENTED** [review]
Good
